docs/RFCS/20200507_partial_indexes.md

```diff
@@ -0,0 +1,318 @@
+
+## Parsing
+
+In order to ensure that predicates are valid (e.g. they result in booleans and
+contain no impure functions), we will use the same logic that validates `CHECK`
+constraints, `sqlbase.SanitizeVarFreeExpr`. The restrictions for `CHECK`
+constraints and partial index predicates are the same.
+
+## Testing Predicate Implication
+
+In order to use a partial index to satisfy a query, the filter expression of the
+query must _imply_ that the partial index predicate is true. If the predicate is
+not provably true, the rows to be returned may not exist in the partial index,
+and it cannot be used.
+
+### Exact matches
+
+First, we will check if any conjuncted-expression in the filter is an exact
+match to the predicate.
+
+For example, consider the filter expression `a > 10 AND b < 100` and the partial
+index predicate `b < 100`. The second conjuncted expression in the filter, `b <
+100`, is an exact match to the predicate `b < 100`. Therefore this filter
+implies this predicate.
+
+We can test for pointer equality to check if the conjuncted-expressions are an
+exact match. The `interner` ensures that identical expressions have the same
+memory address.
+
+### Non-exact matches
+
+There are cases when an expression implies a predicate, but is not an exact
+match.
+
+For example, `a > 10` implies `a > 0` because all values for `a` that satisfy
+`a > 10` also satisfy `a > 0`.
+
+Constraints and constraint sets can be leveraged to help perform implication
+checks. However, they are not a full solution. Constraint sets cannot represent
+a disjunction with different columns on each side.
+
+Consider the following example:
+
+```sql
+CREATE TABLE products (id INT PRIMARY KEY, price INT, units_sold INT, review_count INT)
+CREATE INDEX popular_prds ON t (price) WHERE units_sold > 1000 OR review_count > 100
+```
+
+No constraint can be created for the top-level predicate expression of
+`popular_prds`.
+
+Therefore, constraints alone cannot help us determine that `popular_prds` can be
+scanned to satisfy any of the below queries:
+
+```sql
+SELECT COUNT(id) FROM products WHERE units_sold > 1500 AND price > 100
+
+SELECT COUNT(id) FROM products WHERE review_count > 200 AND price < 100
+
+SELECT COUNT(id) FROM products WHERE (units_sold > 1000 OR review_count > 200) AND price < 100
+```
+
+In order to accommodate for such expressions, we must walk the filter and
+expression trees. At each predicate expression node, we will check if it is
+implied by the filter expression node.
+
+Postgres's [predtest library](https://github.com/postgres/postgres/blob/c9d29775195922136c09cc980bb1b7091bf3d859/src/backend/optimizer/util/predtest.c#L251-L287)
+uses this method to determine if a partial index can be used to satisfy a query.
+The logic Postgres uses for testing implication of conjunctions, disjunctions,
+and "atoms" (anything that is not an `AND` or `OR`) is as follows:
+
+    ("=>" means "implies")
+
+    atom A => atom B if:          A contains B
```

Comment:
Good idea. I was originally thinking of using `constraint.Constraint`, but we only need `constraint.Span` if we determine the columns being operated on are the same. I'll update the spec.

However, I'm just now realizing there are some tricky cases involving `NULL` that we may have to handle without the aid of `constraint.Span`.

For example, imagine these predicates:

1. `a IS NOT NULL`
   * Postgres proves that the following filters imply this:
    * `a IS NOT NULL`
    * `a = 1`
    * `a > 3`
1. `(a, b) IS NULL`
  * Postgres proves that the following filters imply this:
    * `(a, b) IS NULL`
    * `a IS NULL and b IS NULL`
2. `(a, b) IS NOT NULL`
  * Postgres proves that the following filters imply this:
    * `(a, b) IS NOT NULL`
    * `a IS NOT NULL AND b IS NOT NULL`
    * `a = 1 AND b = 2`
    * `a > 5 and b < 7`

For the tuple cases, it might be pragmatic to only handle cases when an atom matches exactly (pointer equality), and leave it at that until we hear from a customer. Heck, it might even be good to push users towards using an equivalent predicate expression that's easier to comprehend, like `a IS NOT NULL AND b IS NOT NULL`.

Also, the PR for fixing `<tuple> IS NULL` logic may make (2) and (3) easier to deal with, if we normalize-away the tuples: https://github.com/cockroachdb/cockroach/pull/48299#pullrequestreview-411200277